### PR TITLE
faster kinematic motion for entities, take 2

### PR DESCRIPTION
**Pull request by AndrewMeadows**
- improve performance in domain with many moving kinematic objects (better, but completely solved)
    - faster kinematic integration (fewer mutex locks, fewer copies)
    - kinematic collision group no longer "touches" static or other kinematic objects
    - don't constantly update broadphase Aabb's of static objects that have not moved
    - faster isNaN() utils
- fix bug for kinematic objects: sudden jump when first set moving
- better Bullet profiler stats
- this time with fewer bugs

### CODE REVIEW

**Comment by AndrewMeadows** [comment]
Hrm... this is not quite ready for testing or review.  I just realized there is another subtle bug in this code.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 51c604b4315a5749675d8461d20f794e60948cd5 available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.dmg

**Comment by AndrewMeadows** [comment]
Ok this is ready to go: a second try at what #7544 was trying to accomplish.  Here are the related asana issues:

https://app.asana.com/0/26623005675174/106408508227390
https://app.asana.com/0/26623005675174/106421247207454

Also this one is related, but not a bug:

https://app.asana.com/0/65049889409317/94635640903820

**Comment by AndrewMeadows** [comment]
Here is the updated testplan:

**performance**
(1) Login and navigate to cellscience domain, wait for overlay buttons to appear near window bottom
(2) Click on the "Inside the cell" button
(3) While you are waiting for all the models to load:
(3A) Enable stats overlay (press the / button, or enable Developer-->Stats menu option)
(3B) Enable detailed stats (Developer-->Timing-->Performance Timer-->Display Timing Details)
(3C) Look for the stat with path /idle/update/physics/stepSimulation
(4) Once the content is loaded look at the value of the /idle/update/physics/stepSimulation stat detail, which will vary over time. Watch the stat for a minute or two and note its average/min/max values.
(5) The average/min/max values of the stat in (4) should be lower in this PR than in the previous release.

Here are the values I was seeing in cellscience for release builds on my linux laptop (2013 MacBook pro, 16GB RAM, GeForce GT 750M):

hifi/master : 4 - 12 msec in stepSimulation, average about 7
this PR: 2 - 9 msec in stepSimulation, average about 4

**kinematic motion still works**
(6) Make entityA (not a sphere!) that is NOT dynamic but nevertheless has a collision shape in the physics simulation (your avatar should be able to bump up against entityA)
(7) Make sure entityA has non-zero linear/angular damping values (so that it will eventually come to rest when it is set moving in step (8))
(8) Give entityA some reasonable velocity (1m/sec along one of the cardinal world axes)
--> entityA should move and slow down to a full stop (note there is a bug that prevents the velocity value in the entity properties from getting zeroed -- not new, we'll fix that in another PR)
(9) Verify that entityA collides correctly where it appears to be
(10) Give entityA some reasonable angular velocity (Note, the angular velocity is still in degrees/sec so a value of 60 should be noticeably fast enough)
--> entityA should turn and eventually rotate to a full stop
(11) Verify that entityA collides correctly for its apparent rotation

**moved static object collides correctly**
(12) Set the velocity of entityA to zero --> it is now "static" in the physics simulation
(13) Change the position of entityA a few meters to one side
(14) Verify that entityA collides correctly for its new apparent position.

**very low gravity**
(15) While entityA is static (not moving)...
(16) Set gravity negative along y aixs but at a very low magnitude (-0.04 or smaller)
(17) Make sure damping it very small or zero (0.1 or smaller)
(18) Give entityA a low positive velocity along the y-axis (on the order of 0.1m)
--> it should rise up and slow down... eventually coming to a complete halt. This is expected: basically we don't support full ballistic motion for objects that have a very small acceleration of gravity.
--> NOTE: once the object has stopped, if you examine the velocity property you may notice that its components are not completely zero. This is a known bug and we'll fix it later.
(19) Try again with a negative acceleration of about -0.09, and give it a small upward velocity (about 0.2)
--> it should rise up, stop, and start to fall again

**dynamic objects coming to rest**
(20) start the dice.js script
(21) create some dice where they can fall to a floor and come to rest
(22) verify that the dice settle down just fine

**Comment by hifi-gustavo** [comment]
Build console output for commit: 95bee53761eb7eacd876750d5e545ee5255d1b0a available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.dmg

**Comment by OmegaHeron** [comment]
Tested Windows 7
Cellscience - inside the cell - physics step time over 2 minutes averages 4.5ms
Cube entity - Linear velocity tests (testing X,Y,Z +/-) - all decay to stop and AV collides with cube while move and  once it comes to full stop.
Cube entity - Rotation tested on X,Y,Z - collides with AV during movement and after coming to rest.
Cube entity - returning to static test - AV collides with cube.

Fail begins at very low gravity test  I have a 1M x 1M cube entity used in tests above.  Set its gravity to -0.04 on Y - set damping to 0 - set Y velocity to 0.1 and it climbs but does not come to rest.  In PR7544 it would climb a bit then stop.  I caught it in edit mode and stopped it once it had risen 20M.

**Comment by AndrewMeadows** [comment]
Thanks @OmegaHeron.  The failure of the low-gravity motion to stop at the peak isn't really a "failure" so much as an expected bug/limitation/tradeoff relating to how we decide to "stop" things that are almost stationary but are not quite.  BTW, what was your frame rate when testing low-gravity?  A slower framerate will tend to make low gravity objects move past their peak -- because at faster framerates the low-velocity status is checked more often and is more likely to trigger, an alternative way to think about it is that at low framerates the object is more likely to accelerate through the "it's probably stopped" threshold logic.

**Comment by OmegaHeron** [comment]
If I unlocked vsync > 100, so 60 for this test.  It's simply a single cube platform to stand on in open space where I test things like this.

I think I wasn't clear tho - in the previous PR7444 the very low gravity test did stop - cube would climb some then come to rest (yes - edit props showed it not at 0.0 velocity per expectations), but it rose - slowed and came to rest.  In this case the positive vertical velocity remains constant and if you don't edit to stop it will climb essentially forever showing no apparent change in velocity over time.  If that's the expected - cool.

**Comment by AndrewMeadows** [comment]
Oh... no that sounds like a bug, or something.  I will verify.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1d80494d503a4cd0e0726839407392c4691106db available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7590.dmg

**Comment by AndrewMeadows** [comment]
Ok, I've fixed the final bug and then a few other problems I noticed along the way.  This PR is getting big and complicated but I stand by it all -- we need this work merged eventually.  If anyone has qualms then I can attempt to sooth them by breaking this up into two or three independent PR's.

**Comment by OmegaHeron** [comment]
Testing Windows 7

Cellscience - Same as previous tests (better performance on step time).

Assigning linear and rotational movement to cube entity (tested application of all on X/Y/Z) AV can collide with cube during movement and once it comes to rest.
Back to static object - AV collides with cube.

Very low gravity - all tests yield results indicated in testing procedure.

Dice Test - threw dice - rolled on floor (a cube entity) and came to rest.  Tested against a mesh object with compound shape - same as cube entity.  

QA Pass

**Comment by AndrewMeadows** [comment]
@ZappoMan was asking for a benchmark for evaluating the risk/profit balance for merging this work sooner vs later.  I've created a script below that can be used to measure performance changes between this PR and other versions of interface:

```
//
//  kinematicBenchmark.js -- stress test for performance bugs in kinematic motion
//
//  Created by Andrew Meadows 2016.04.07
//  Copyright 2016 High Fidelity, Inc.
//
//  Distributed under the Apache License, Version 2.0.
//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
//

// Here is repro recipe:
//
// (1) Enable stats, with details:
// (a) Developer --> Stats
// (b) Developer --> Timing --> Performance Timer --> Display Timing Details
//
// (2) Navigate to about <50, 50, 50> and look toward the world's origin
// (3) Run this script
// (4) Press 't' to start the objects moving.
// (5) Watch the stats in /idle/update/ as the entities move about.

// This script creates a number of non-dynamic entities randomly about a sphere.
// When the 't' key is pressed the entities start moving in a linear kinematic
// path either toward the center of the sphere or away depending on the test mode.
//
// There were two main performance problems addressed hence this test script has two
// modes: one to test the performance problem in Bullet and another to test that of
// our own custom kinematic motion calculations (entity-math).  Uncomment the mode
// of your choice and then reload the script:

//var TEST_MODE = "bullet-aabb-optimizations";
var TEST_MODE = "hifi-kinematic-optimizations";

// for the visual model we use the six-sided die, because it is relatively inexpensive to render
var HIFI_PUBLIC_BUCKET = "http://s3.amazonaws.com/hifi-public/";
var MODEL =  HIFI_PUBLIC_BUCKET + "models/props/Dice/goldDie.fbx"

// for the collision shape we use the tiltmaze because it hast a lot of vertices and produces
// a rather expensive compound shape and will stress the physics Aabb updates
var COLLISION_HULL = "http://hifi-content.s3.amazonaws.com/DomainContent/Home/tiltMaze/newmaze_tex-3.obj";

var SPHERE_CENTER = {x: 0, y: 30, z: 0};
var SPHERE_RADIUS = 20;
var PERIOD = 10.0;
var BASE_DIMENSION = 2;

var NUM_ENTITIES = 10;
var VELOCITY_SIGN = 1.0;

if (TEST_MODE == "hifi-kinematic-optimizations") {
    NUM_ENTITIES = 100;
    VELOCITY_SIGN = -1.0;
}

KinematicObject = function() {
    this.entityId = null;
    this.startPosition = {x: 0, y: 0, z: 0};
    this.endPosition = {x: 0, y: 0, z: 0};
    this.period = PERIOD;
    this.velocity = {x: 0, y: 0, z: 0};
};

var objectCount = 0;

KinematicObject.prototype.createEntity = function() {
    if (this.entityId == null) {
        var dimension = BASE_DIMENSION * (1.0 + Math.random());
        if (TEST_MODE == "bullet-aabb-optimizations") {
            this.entityId = Entities.addEntity( {
                type: "Model",
                modelURL: MODEL,
                compoundShapeURL: COLLISION_HULL,
                name: "obj" + objectCount,
                dimensions: {x: dimension, y: dimension, z: dimension},
                position: this.startPosition,
                velocity: {x: 0, y: 0, z: 0},
                gravity: {x: 0, y: 0, z: 0},
                angularVelocity: {x: 0, y: 0, z: 0},
                color: { red: Math.random() * 255, green: Math.random() * 255, blue: Math.random() * 255 },
                dynamic: false,
                collisionless: false,
                damping: 0,
                lifetime: 300
            } );
        } else {
            this.entityId = Entities.addEntity( {
                type: "Box",
                name: "obj" + objectCount,
                dimensions: {x: dimension, y: dimension, z: dimension},
                position: this.startPosition,
                velocity: {x: 0, y: 0, z: 0},
                gravity: {x: 0, y: 0, z: 0},
                angularVelocity: {x: 0, y: 0, z: 0},
                color: { red: Math.random() * 255, green: Math.random() * 255, blue: Math.random() * 255 },
                dynamic: false,
                collisionless: false,
                damping: 0,
                lifetime: 300
            } );
        }
        objectCount++;
    }
}

KinematicObject.prototype.deleteEntity = function() {
    if (this.entityId != null) {
        Entities.deleteEntity(this.entityId);
        this.entityId = null;
    }
}

KinematicObject.prototype.start = function() {
    if (this.entityId != null) {
        var travel = {
            x: this.endPosition.x - this.startPosition.x,
            y: this.endPosition.y - this.startPosition.y,
            z: this.endPosition.z - this.startPosition.z,
        };
        var period = this.period;
        this.velocity = {
            x: VELOCITY_SIGN * travel.x / period,
            y: VELOCITY_SIGN * travel.y / period,
            z: VELOCITY_SIGN * travel.z / period
        }
        var properties = {
            position: this.startPosition,
            velocity: this.velocity
        };
        Entities.editEntity(this.entityId, properties);
    }
}

KinematicObject.prototype.stop = function() {
    if (this.entityId != null) {
        var properties = {
            position: this.endPosition,
            velocity: {x: 0, y: 0, z: 0}
        };
        Entities.editEntity(this.entityId, properties);
    }
}

function randomDirection() {
    var x = Math.random() - 0.5;
    var y = Math.random() - 0.5;
    var z = Math.random() - 0.5;
    var length = Math.sqrt(x * x + y * y + z * z);
    return {x: x / length, y: y / length, z: z / length};
}

var objects = [];
var i = 0;
for (i = 0; i < NUM_ENTITIES; i++) {
    var obj = new KinematicObject;
    obj.createEntity();
    var direction = randomDirection();
    obj.startPosition = {
        x: SPHERE_CENTER.x - SPHERE_RADIUS * direction.x,
        y: SPHERE_CENTER.y - SPHERE_RADIUS * direction.y,
        z: SPHERE_CENTER.z - SPHERE_RADIUS * direction.z
    };
    obj.endPosition = {
        x: SPHERE_CENTER.x + SPHERE_RADIUS * direction.x,
        y: SPHERE_CENTER.y + SPHERE_RADIUS * direction.y,
        z: SPHERE_CENTER.z + SPHERE_RADIUS * direction.z
    };
    obj.stop();
    objects.push(obj);
}

var running = false;

function stopEverything() {
    if (running) {
        running = false;
        for (i = 0; i < NUM_ENTITIES; i++) {
            var obj = objects[i];
            obj.stop();
        }
    }
}

function pressEvent(event) {
    if (event.text == "t") {
        running = !running;
        if (running) {
            Script.setTimeout(stopEverything, 1000 * PERIOD);
            for (i = 0; i < NUM_ENTITIES; i++) {
                var obj = objects[i];
                obj.start();
            }
        } else {
            for (i = 0; i < NUM_ENTITIES; i++) {
                var obj = objects[i];
                obj.stop();
            }
        }
    }
}

function tearDown() {
    var i = 0;
    for (i = 0; i < NUM_ENTITIES; i++) {
        var obj = objects[i];
        obj.deleteEntity();
    }
}

Controller.keyPressEvent.connect(pressEvent);
Script.scriptEnding.connect(tearDown);
```

**Comment by AndrewMeadows** [comment]
Normally I would let these changes slide under the radar, but for those interested:

**First kinematic step was incorrect:**
(1) Create an entity
(2) Give it 1m/sec velocity in some direction
--> On previous releases: the entity would incorrectly measure the first timeStep, clamp it to 1.0sec, and pop the object forward.  The size of the pop is directly proportional to the velocity.  This also happens for angular velocity.
--> Now fixed.

**Kinematic motion was not owned in distributed simulation**
(3) Create a kinematic object (non-dynamic, but with non-zero velocity)
--> On previous releases: the script operation that sets the velocity causes a bid for ownership, but the ownership is promptly lost and the motion is computed on interface and entity-server independently, which means they could end up in different positions.  Typically this is not a problem and the positions will only be a few mm or cm off, but for slow trajectories at the peak of their ballistic motion it is theoretically possible for one simulation to decide to terminate its motion (clamp velocities to zero) while the other one does not --> all participating interfaces and entity-server could end up in very different places.  Eventually the entity-server would resend the data and the object would pop on wayward interfaces.
--> Now the interface that starts the kinematic motion will maintain ownership until it comes to rest, and will send a final update thereby synchronizing all participants.

**Comment by AndrewMeadows** [comment]
Could someone run the benchmark script above and report back.  Also, please verify the "pop" in the kinematic motion is fixed.

**Comment by OmegaHeron** [comment]
tldr version - this can be dramatically **faster**.  Apologies for not being able to come up with a better display format for data - not enough coffee yet even by East Coast time zone standards.

Tested in almost empty domain space - GTX650M i7 CPU 3GHz

**TEST_MODE = "hifi-kinematic-optimizations"**

_Before running script for baseline_

|Windows Build 4657   | This PR  |
|---|---|
| physics **0.07** physics/stepSimulation **0.025**   | physics **0.06** physics/stepSimulation **0.023**  |
| Render Rate 60 Present Rate 60 Sim Rate 60  | Render Rate 60 Present Rate 60 Sim Rate 60  |

_Script running, objects in view - not yet moving._

|Windows Build 4657   | This PR  |
|---|---|
| physics **0.09** physics/stepSimulation **0.05**   | physics **0.07** physics/stepSimulation **0.026**  |
| Render Rate 60 Present Rate 60 Sim Rate 60  | Render Rate 60 Present Rate 60 Sim Rate 60  |

_In Motion_

|Windows Build 4657   | This PR  |
|---|---|
| physics **2.5(max)** physics/stepSimulation **0.45(max)**   | physics **0.7(max)** physics/stepSimulation **0.45(max)**  |
| Render Rate 60 Present Rate 60 Sim Rate 60 (Stable)  | Render Rate 60 Present Rate 60 Sim Rate 60  |

**TEST_MODE = "bullet-aabb-optimizations"**

_Before running script for baseline_

|Windows Build 4657   | This PR  |
|---|---|
| physics **0.07** physics/stepSimulation **0.025**   | physics **0.06** physics/stepSimulation **0.023**  |
| Render Rate 60 Present Rate 60 Sim Rate 60  | Render Rate 60 Present Rate 60 Sim Rate 60  |

_Script running, objects in view - not yet moving._

|Windows Build 4657   | This PR  |
|---|---|
| physics **0.07** physics/stepSimulation **0.03**   | physics **0.07** physics/stepSimulation **0.03**  |
| Render Rate 60 Present Rate 60 Sim Rate 60  | Render Rate 60 Present Rate 60 Sim Rate 60  |

_In Motion_

|Windows Build 4657   | This PR  |
|---|---|
| physics **44.0(max)** physics/stepSimulation **44.0(max)**   | physics **0.26(max)** physics/stepSimulation **0.08(max)**  |
| Render Rate 60 down to 8 Present Rate 60 Sim Rate 60 down to 8  | Render Rate 60 Present Rate 60 Sim Rate 60 (Stable)  |

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9f43187ebf783b2af1e0a1d5c0081b218019d4d8 available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0403be0ca6c2f2bd6b46dd2b7498e71b3df7e1ab available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.dmg

**Comment by Nex-Pro** [comment]
This was extremely slow to install .... I'm still waiting for the installer to finish

**Comment by cozza13** [comment]
@Nex-Pro , this does not need QA at the moment. It is marked approved

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70673cc379bc143552e7285e466f3ae3ab903754 available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.dmg

**Comment by AndrewMeadows** [comment]
rebuilding mac platform
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 70673cc379bc143552e7285e466f3ae3ab903754 available at: https://gist.github.com/cc7238d38119f15b178df6222f807c3e<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7590.dmg

**Comment by AndrewMeadows** [comment]
There is a reported bug with this PR: ballistic dynamic objects jitter and glitch through their path.

**Comment by AndrewMeadows** [comment]
I'm closing this PR.  I will split up the development into two parts:

(1) Bullet Aabb and collision group changes
(2) EntityItem::simulateKinematicMotion() changes

Part (1) is already up as #7643
